Fixed consistency of behavior boolean properties

koba04 opened the pull request:
The behavior of DOM Component boolean property is different by the component lifecycle (create or update).

```js
var Button = React.createClass({
    getInitialState: function() {
        return {
            disabled: true
        };
    },
    render: function() {
        return <div><button disabled={this.state.disabled}>click</button></div>;
    }
});
 
var component = React.render(<Button />, document.getElementById('container'));
console.log(React.findDOMNode(component).innerHTML);
// <button disabled="" data-reactid=".0.0">click</button>

component.setState({disabled: false});
console.log(React.findDOMNode(component).innerHTML);

component.setState({disabled: true});
console.log(React.findDOMNode(component).innerHTML);
// <button data-reactid=".0.0" disabled="true">click</button>
```

https://jsfiddle.net/koba04/arehgdsq/

I think it should be the same behavior regardless of the component lifecycle.
This PR applies the DOM operation in the creating component process to the updating process.

zpao commented on src/renderers/dom/shared/DOMPropertyOperations.js:
```diff
@@ -174,7 +174,14 @@ var DOMPropertyOperations = {
           node.setAttributeNS(namespace, attributeName, '' + value);
         } else {
-          node.setAttribute(attributeName, '' + value);
+          if (propertyInfo.hasBooleanValue ||
```
Can make the above an `else if` and then the `else` stays the same (and we avoid nested conditional logic)

zpao commented:
Yea, this makes to me. Just the one comment.

koba04 commented:
Thank you for your review!
I've fixed it.
In addition to that, I removed the unnecessary check for `value`.
`value` is always evaluated as `true`.

https://github.com/facebook/react/blob/master/src%2Frenderers%2Fdom%2Fclient%2FReactDOMIDOperations.js#L61-L65

zpao commented:
> `value` is always evaluated as `true`

I'm pretty sure that's not correct. Eg, `checked={false}` Though maybe that doesn't get down this codepath?

koba04 commented:
@zpao 
The mentioned URL was not enough...

> value is always evaluated as true

But that seems to be correct.
`false` value doesn't get down this codepath because it matches the `shouldIgnoreValue`. 

```js
function shouldIgnoreValue(propertyInfo, value) {
  return value == null ||
    (propertyInfo.hasBooleanValue && !value) ||
    (propertyInfo.hasNumericValue && isNaN(value)) ||
    (propertyInfo.hasPositiveNumericValue && (value < 1)) ||
    (propertyInfo.hasOverloadedBooleanValue && value === false);
}
```

https://github.com/facebook/react/blob/master/src/renderers/dom/shared/DOMPropertyOperations.js#L45-L51

```js
      } else if (shouldIgnoreValue(propertyInfo, value)) {
        this.deleteValueForProperty(node, name);
      } else if (propertyInfo.mustUseAttribute) {
```

https://github.com/facebook/react/blob/master/src/renderers/dom/shared/DOMPropertyOperations.js#L167-L169






And also `checked` value doesn't get down this codepath because it isn't `MUST_USE_ATTRIBUTE`.

koba04 commented:
@zpao 
Is there anything else to do?

zpao commented:
:thumbsup: Thanks!

koba04 commented:
Thank you! :tada:
